src/models/streaming/BufferedDuplex.ts

```diff
@@ -0,0 +1,43 @@
+import { Duplex, PassThrough } from 'stream';
+
+export default class BufferedDuplex extends Duplex {
+    input: PassThrough;
+    output: PassThrough;
+    outputHandlersSetup: boolean;
+    constructor (input: PassThrough, output: PassThrough, options?: any) {
+        super(options);
+        this.input = input;
+        this.output = output;
+        this.outputHandlersSetup = false;
+    }
+
+    _write (chunk: any, encoding: any, callback: any) : boolean {
+        return this.input.write(chunk, encoding, callback);
+    }
+
+    _read (size: any): any {
+        if (!this.outputHandlersSetup){
+            return this.setUpOutputHandlersAndRead(size);
+        }
+        return this.readOutput(size);
+    }
+
+    setUpOutputHandlersAndRead (size : any): void {
+        var self = this;
+        self.outputHandlersSetup = true;
```

Comment:
minor, to follow the flow of the `outputHandlersSetup` I would move this assignment to after the handlers have been set